### PR TITLE
feat(pipelines): changes needed to support other engines

### PR DESCRIPTION
**Pull request by eladb**
This PR includes changes needed in the Pipelines library that can help with implementation of other pipeline engines. Namely, these changes were identified during work to implement a GitHub workflows engine.

1. Add support in `PipelineStructure` for publishing the CFN template of a stack to the assets bucket. This is enabled by setting `publishTemplate: true`.
2. Add support in `PipelineStructure` for omitting the "prepare" step in case the engine only wants a single step for each stack deployment. This is enabled by setting `prepareStep: false`.
3. Add `graphNode.uniqueId` which returns a graph-wide unique ID of the node.
4. Add `graphNode.allDeps` which returns a union of the node's dependencies and the dependencies of all its parents.
5. Renamed `PipelineStructure` to `PipelineGraph`.

Additionally, in order to allow engine providers to leverage these helpers, I've moved all these types under `lib/helpers-internal` so they can be imported through a barrel import: `import { PipelineGraph } from '@aws-cdk/pipelines/helpers-internal'`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/pipelines/lib/blueprint/stack-deployment.ts [hunk 1]
```diff
@@ -136,9 +166,15 @@ function extractStackAssets(stackArtifact: cxapi.CloudFormationStackArtifact): S
...
       });
     }
   }
 
   return ret;
+}
+
+function s3UrlFromUri(uri: string) {
+  const url = parseUrl(uri);
+  return `https://${url.hostname}.s3.amazonaws.com${url.path}`;
```
Comment (eladb) [inline packages/@aws-cdk/pipelines/lib/blueprint/stack-deployment.ts:179]: Do we need this to be region-aware?

File: packages/@aws-cdk/aws-lambda/package.json [hunk 2]
```diff
@@ -183,7 +183,8 @@
   },
   "nozem": {
     "ostools": [
-      "docker"
+      "docker",
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda/package.json:186]: @rix0rrr I needed this to use nozem. WDYT?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-lambda/package.json:186]: You need a newer version of Nozem. A newer version will do this automatically.

**Comment by rix0rrr** [comment]
> Add support in PipelineStructure for publishing the CFN template of a stack to the assets bucket. This is enabled by setting publishTemplate: true.

Wait up -- you shouldn't need to do this. The `DefaultSynthesizer` already adds the stack template as an asset, and so it will be in the asset manifest and be uploaded as part of the regular asset publishing step.

**Comment by rix0rrr** [comment]
> Consider if we want to move this from "test" to a public API (maybe with a Beta1 suffix).

If we want to do this properly we should be able to publish it as a plugin package on NPM, no? (That would prove that other people would be able to do the same)

**Review by rix0rrr — COMMENTED** [review]
Where do you need `CreateStack` again?

**Comment by eladb** [comment]
> > Add support in PipelineStructure for publishing the CFN template of a stack to the assets bucket. This is enabled by setting publishTemplate: true.
> 
> 
> 
> Wait up -- you shouldn't need to do this. The `DefaultSynthesizer` already adds the stack template as an asset, and so it will be in the asset manifest and be uploaded as part of the regular asset publishing step.

This is about considering the template asset in the pipeline graph (not about adding another asset to the assembly). Without this, the PipelineStructure class used to simply filter out this asset.

**Comment by eladb** [comment]
> Where do you need `CreateStack` again?

It's needed by the GitHub action I am using.

**Comment by eladb** [comment]
> > Consider if we want to move this from "test" to a public API (maybe with a Beta1 suffix).
> 
> 
> 
> If we want to do this properly we should be able to publish it as a plugin package on NPM, no? (That would prove that other people would be able to do the same)

Good point. Definitely doable. I guess I can extract this already to a separate package.

File: packages/@aws-cdk/aws-lambda/package.json (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-lambda/package.json]: Reverted

File: packages/@aws-cdk/pipelines/lib/blueprint/stack-deployment.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/pipelines/lib/blueprint/stack-deployment.ts:179]: Yes

**Review by rix0rrr — APPROVED** [review]

File: packages/@aws-cdk/pipelines/lib/blueprint/stack-deployment.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/pipelines/lib/blueprint/stack-deployment.ts]: Done

**Pull request merged**